Cleanup some dialog and popover focus steps to bring in line with the spec

beidson opened the pull request:
#### d131c3505c51e1c9bb14b75420aa43526c179742
<pre>
Cleanup some dialog and popover focus steps to bring in line with the spec
<a href="https://rdar.apple.com/148533180">rdar://148533180</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291011">https://bugs.webkit.org/show_bug.cgi?id=291011</a>

Reviewed by Ryosuke Niwa.

These autofocus-related steps were subtly wrong in how they chose the top document on which to
clearAutofocusCandidates.

This patch fixes that up. With the great side benefit of removing some site isolation concerns.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::sameOriginTopLevelTraversable const): This precise concept is useful in at least
    (3) call sites in autofocus code, and is likely to also be useful in other places later.
* Source/WebCore/dom/Document.h:

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::insertedIntoAncestor): Adopt sameOriginTopLevelTraversable.

* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps): Adopt sameOriginTopLevelTraversable.

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps): Adopt sameOriginTopLevelTraversable.

Canonical link: <a href="https://commits.webkit.org/293261@main">https://commits.webkit.org/293261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d608f3a2da9424f4b4faa501f3e3fc00c9e16357

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98341 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17972 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8200 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103461 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48870 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100386 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18264 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26423 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74862 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32024 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101345 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13836 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88819 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55221 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13618 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6780 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48312 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105836 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25428 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83842 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25801 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85021 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83313 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21050 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27950 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5625 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19076 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25386 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25206 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26781 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7696,6 +7696,26 @@ Document* Document::mainFrameDocument() const
     return document;
 }
 
+RefPtr<Document> Document::sameOriginTopLevelTraversable() const
```
I'd call this `mainFrameDocumentIfSameOrigin` or something.
The current function name makes me think that it would return the highest ancestor frames which is of the same origin, not return nullptr when the main frame's origin differs.

nt1m replied:
I think the current name matches the spec, which I think is a nice benefit

beidson replied:
Correct - This name is purposefully derived from spec language, and as such when to use it should (hopefully) be very obv.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7696,6 +7696,26 @@ Document* Document::mainFrameDocument() const
+    while (Frame* parent = topLevelAncestorFrame->tree().parent())
+        topLevelAncestorFrame = parent;
+
+    RefPtr<LocalFrame> localTopAncestor = dynamicDowncast<LocalFrame>(topLevelAncestorFrame);
```
```suggestion
    RefPtr localTopAncestor = dynamicDowncast<LocalFrame>(topLevelAncestorFrame);
```

beidson committed:
Cleanup some dialog and popover focus steps to bring in line with the spec
rdar://148533180
https://bugs.webkit.org/show_bug.cgi?id=291011

Reviewed by Ryosuke Niwa.

These autofocus-related steps were subtly wrong in how they chose the top document on which to
clearAutofocusCandidates.

This patch fixes that up. With the great side benefit of removing some site isolation concerns.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::sameOriginTopLevelTraversable const): This precise concept is useful in at least
    (3) call sites in autofocus code, and is likely to also be useful in other places later.
* Source/WebCore/dom/Document.h:

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::insertedIntoAncestor): Adopt sameOriginTopLevelTraversable.

* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps): Adopt sameOriginTopLevelTraversable.

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps): Adopt sameOriginTopLevelTraversable.

Canonical link: https://commits.webkit.org/293261@main
